File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* solution = new ListNode;
        ListNode* move1 = l1;
        ListNode* move2 = l2;
        ListNode* move3 = solution;
        
        for (int carry = 0; move1 != nullptr || move2 != nullptr || carry == 1;) {
            int val1 = 0, val2 = 0;
            if(move1 != nullptr) val1 = move1->val;
            if(move2 != nullptr) val2 = move2->val;
            move3->val = (val1 + val2 + carry)%10;
            if ((val1 + val2 + carry) > 9)
                carry = 1;
            else 
                carry = 0;
            if(move1 != nullptr) move1 = move1->next;
            if(move2 != nullptr) move2 = move2->next;
            if (move1 != nullptr || move2 != nullptr || carry == 1) move3->next = new ListNode;
            if (move1 != nullptr || move2 != nullptr || carry == 1) move3 = move3->next;
        }
        
        return solution;   
    }
    
};

File: 0002-add-two-numbers/0002-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    // brute force
    var arr1 = [];
    var arr2 = [];
    while (l1 !== null) {
        arr1.push(l1.val);
        l1 = l1.next;
    }
    while (l2 !== null) {
        arr2.push(l2.val);
        l2 = l2.next;
    }
    var res = addTwoArrays(arr1, arr2);
    if (res.length === 0) {
        return null;
    }
    var head = new ListNode(res[0]);
    var prev = head;
    for (var i = 1; i < res.length; i++) {
        var node = new ListNode(res[i]);
        prev.next = node;
        prev = node;
    }
    return head;
};

var addTwoArrays = (a1, a2) => {
    // return result array in reverse order
    var result = [];
    if (a1.length > a2.length) {
        var carry = 0;
        for (var i = 0; i < a1.length; i++) {
            if (a2[i] !== undefined) {
                result.push((a1[i] + a2[i] + carry) % 10);
                carry = a1[i] + a2[i] + carry >= 10 ? 1 : 0;
            } else {
                result.push((a1[i] + carry) % 10);
                carry = a1[i] + carry >= 10 ? 1 : 0;
            }
        }
        if (carry === 1) {
            result.push(1);
        }
    } else {
        var carry = 0;
        for (var i = 0; i < a2.length; i++) {
            if (a1[i] !== undefined) {
                result.push((a1[i] + a2[i] + carry) % 10);
                carry = a1[i] + a2[i] + carry >= 10 ? 1 : 0;
            } else {
                result.push((a2[i] + carry) % 10);
                carry = a2[i] + carry >= 10 ? 1 : 0;
            }
        }
        if (carry === 1) {
            result.push(1);
        }
    }
    return result;
}